File: lib/glg/SimpleViewer/SingleThreaded/Java/GlgViewer.java
import java.awt.*;
import java.awt.event.*;
import javax.swing.*;
import java.lang.reflect.*;
import java.util.ArrayList;
import com.genlogic.*;

//////////////////////////////////////////////////////////////////////////
public class GlgViewer extends GlgJBean implements ActionListener
{
   /* Set to true to use simulated data for animation. 
      Set to false to use live real-time data.
      May be overriden by command line option -random-data or -live-data in
      a standalone program, or RandomData applet parameter if used
      inside an applet. 
   */
   public boolean RANDOM_DATA = true;

   // Parent container.
   SimpleViewer Parent;

   // Dynamically created array of tag records.
   ArrayList<GlgTagRecord> TagRecordArray = new ArrayList<GlgTagRecord>();
   
   // Number of tags records in the TagRecordArray.
   public int NumTags = 0;

   // DataFeed object is used to supply data for animation.
   DataFeedInterface DataFeed;

   Timer timer = null;
   int UpdateInterval = 100; // update interval in msec.
   boolean IsReady = false;

   // Used by DataFeed to return data values.
   DataPoint d_data_point = new DataPoint( GlgObject.D );
   DataPoint s_data_point = new DataPoint( GlgObject.S );

   //////////////////////////////////////////////////////////////////////////
   public GlgViewer( SimpleViewer parent )
   {
      super();
      Parent = parent;
   }
      
   //////////////////////////////////////////////////////////////////////
   // StartUpdates() creates a timer to perform periodic updates.
   // The timer invokes the bean's UpdateGraphProc() method to update
   // drawing's resoures with new data values.
   /////////////////////////////////////////////////////////////////////
   public void StartUpdates()
   {
      if( timer == null )
      {
         timer = new Timer( UpdateInterval, this );
         timer.setRepeats( true );
         timer.start();
      }
   }

   ///////////////////////////////////////////////////////////////////////
   // StopUpdate() method stops periodic updates
   ///////////////////////////////////////////////////////////////////////
   public void StopUpdates()
   {
      if( timer != null )
      {
         timer.stop();
         timer = null;
      }
   }
   
   //////////////////////////////////////////////////////////////////////////
   public void LoadDrawing( String drawing_filename )
   {
      if( drawing_filename == null )
      {
         System.out.println( "No drawing file: drawing loading failed." );
         return;
      }

      SetDrawingName( drawing_filename );
   }

   //////////////////////////////////////////////////////////////////////
   public void HCallback( GlgObject viewport )
   {
      AddDataFeed();
   }

   //////////////////////////////////////////////////////////////////////
   // VCallback() is invoked after the drawing is loaded and setup, 
   // but before it is drawn for the first time. 
   /////////////////////////////////////////////////////////////////////
   public void VCallback( GlgObject viewport )
   {
      /* Clear TagRecordArray, in case the drawing
         is replaced in the GlgViewer, so that the previous 
         TagRecordArray is cleared before the new one gets
         created.
      */
      DeleteTagRecords();

      // Build TagRecordArray, a list of glg tag records.
      CreateTagRecords( viewport );
   }
   
   ////////////////////////////////////////////////////////////////////////
   // ReadyCallback is invoked after the drawing is loaded, setup and 
   // initially drawn.
   ////////////////////////////////////////////////////////////////////////
   public void ReadyCallback( GlgObject viewport )
   {
      IsReady = true;
   }

   //////////////////////////////////////////////////////////////////////////
   // Add DataFeed object to a glg_bean.
   //////////////////////////////////////////////////////////////////////////
   public void AddDataFeed()
   {
      if( RANDOM_DATA )
      {
        DataFeed = new DemoDataFeed( this );
        System.out.println( "Using random demo DataFeed." );
      }
      else
      {
        DataFeed = new LiveDataFeed( this );
        System.out.println( "Using live DataFeed." );
      }
   }

   /////////////////////////////////////////////////////////////////////
   // Create and populate TagRecordArray
   /////////////////////////////////////////////////////////////////////
   public void CreateTagRecords( GlgObject viewport )
   {
      GlgObject tag_obj;
      String tag_source;
      String tag_name;
      String tag_comment;
      int tag_access_type;
      GlgObject tag_list;

      /* Retrieve a tag list from the drawing. Include tags with unique
         tag sources.
      */
      tag_list = viewport.CreateTagList( /*unique tag sources*/ true );
      if( tag_list == null )
        return;
      
      int size = tag_list.GetSize();
      if( size == 0 )
        return; // no tags found 
            
      /* Create an array of tag records by traversing the tag list
         and retrieving information from each tag object
         in the list.
      */
      for( int i=0; i<size; ++i )
      {
         tag_obj = (GlgObject) tag_list.GetElement( i );
         tag_source = tag_obj.GetSResource( "TagSource" );
         tag_name = tag_obj.GetSResource( "TagName" );
         tag_comment = tag_obj.GetSResource( "TagComment" );
         
         // Skip undefined tags.
         if( tag_source == null || tag_source.equals("") || 
             tag_source.equals("unset") )
           continue;
         
         tag_access_type = tag_obj.GetDResource( "TagAccessType" ).intValue();
         
         // Handle special tags according to their tag access type.
         switch( tag_access_type )
         {
          case GlgObject.OUTPUT_TAG: continue;   // Skip OUTPUT tags.
            
          case GlgObject.INIT_ONLY_TAG:
            /* Initialize INIT_ONLY tags; do not include them in 
               TagRecordArray, so that they will not be subject to 
               the continous data animation. 
            */
            InitTag( tag_source );
            continue;
          
          default: break;
         }
         
         // Tag source is valid, add a new tag record to TagRecordArray.
         GlgTagRecord tag_record = new GlgTagRecord();
         tag_record.tag_obj = tag_obj;
         tag_record.tag_source = tag_source;
         tag_record.data_type = tag_obj.GetDResource( "DataType" ).intValue();
         
         /* For further performance optimization, set if_changed=true which 
            will push the value into the tag only if the value has changed. 
            Set if_changed = false for the tags assigned to the entry 
            points of a chart object.
         */
         if( tag_comment != null && tag_comment.contains("Chart") )
           tag_record.if_changed = false;
         else
           tag_record.if_changed = true;      
         
         // Add a valid tag record to the list.
         TagRecordArray.add( tag_record );
      }
      
      // Store number of tag records in TagRecordArray.
      NumTags = TagRecordArray.size();
   }
   
   //////////////////////////////////////////////////////////////////////////
   // Clear TagRecordArray
   //////////////////////////////////////////////////////////////////////////
   public void DeleteTagRecords()
   {
      if( TagRecordArray.size() > 0 )
        TagRecordArray.clear();

      NumTags = 0;
   }
   
   //////////////////////////////////////////////////////////////////////////
   // Initialize tag with a given tag source.
   //////////////////////////////////////////////////////////////////////////
   public void InitTag( String tag_source )
   {
      double value = -1.0;
      
      if( RANDOM_DATA )
      {
         // Initialize "State" tag.
         if( tag_source.equals( "State" ) )
           value = 1.0;
      }
      else
      {
         // Place custom code here to set value as needed.
      }
      
      // Set the tag value in the drawing.
      SetDTag( tag_source, value, false );
   }

   ///////////////////////////////////////////////////////////////////////
   // timer's ActionListener method to be invoked periodically 
   //////////////////////////////////////////////////////////////////////
   public void actionPerformed( ActionEvent e )
   {
      UpdateDrawing();
   }
   
   //////////////////////////////////////////////////////////////////////////
   // Update the drawing with new dynamic data values.
   //////////////////////////////////////////////////////////////////////////
   public void UpdateDrawing()
   {
      if( !IsReady )
        return;
      
      /* For each tag in the TagRecordArray, obtain a new data value using 
         DataFeed object and push a new value into the graphics.  
      */
      for( int i = 0; i<NumTags; ++i )
      {
         GlgTagRecord tag_record = TagRecordArray.get(i);
         switch( tag_record.data_type )
         {
          case GlgObject.D:
            if( !DataFeed.ReadDValue( tag_record, d_data_point ) )
            {
               System.out.println( "Failed to read data value for " + 
                                   tag_record.tag_source );
               continue;
            }

            // Push new data value into the graphics.
            SetDTag( tag_record.tag_source,  d_data_point.d_value, 
                     tag_record.if_changed );
            break;

          case GlgObject.S:
            if( !DataFeed.ReadSValue( tag_record, s_data_point ) )
            {
               System.out.println( "Failed to read data value for " + 
                                   tag_record.tag_source );
               continue;
            }

            // Push new data value into the graphics.
            SetSTag( tag_record.tag_source, s_data_point.s_value, 
                     tag_record.if_changed );
            break;

          case GlgObject.G:      // Not used in this example.
          default:
            break;
         }
      }
      
      // Update the display
      Update();
   }
   
   ///////////////////////////////////////////////////////////////////////
   // Handle user interaction.
   ///////////////////////////////////////////////////////////////////////
   public void InputCallback( GlgObject viewport, GlgObject message_obj )
   {
      String
        origin,
        format,
        action,
        subaction;
      
      origin = message_obj.GetSResource( "Origin" );
      format = message_obj.GetSResource( "Format" );
      action = message_obj.GetSResource( "Action" );
      subaction = message_obj.GetSResource( "SubAction" );
      
      // Handle window closing if run stand-alone. 
      if( format.equals( "Window" ) && action.equals( "DeleteWindow" ) )
        System.exit( 0 );
      
      // Handle button events based on the button names.
      if( format.equals("Button") )
      {
         if( !action.equals("Activate") && !action.equals("ValueChanged") )
           return;
         
         if( action.equals("Activate") )  //Push button event
         {
            if( origin.equals("QuitButton") )
            {  
               if( Parent.IsStandalone )
                 System.exit( 0 );
            }
            else if( origin.equals("StartButton") )
            {
               if ( RANDOM_DATA )
               {
                  DataFeed.WriteDValue( "State", 1.0 );
                  
                  // Start/Resume dynamic updates.
                  StartUpdates();
               }
               // Place custom code here as needed.
            }
            else if( origin.equals("StopButton") )
            {
               if ( RANDOM_DATA )
               {
                  DataFeed.WriteDValue( "State", 0.0 );
                  
                  // Stop dynamic updates.
                  StopUpdates();
               }
               else
                 // Place custom code here as needed.
                 ; 
            }
         }
         else if( action.equals("ValueChanged") ) // Toggle button
         {
            double state = message_obj.GetDResource( "OnState" ).doubleValue();

            // Place code here to handle events from a toggle button
            // and write a new value to a given tag_source.
            // DataFeed.WriteDValue( (tag_source, state );
         }
         
         // Make changes visible.
         viewport.Update();
      }
      else if( format.equals( "Timer" ) )
      {
         // Refresh the display for the objects with integrated Timer dynamics.
         viewport.Update();
      }
   }
}

File: lib/glg/SimpleViewer/SingleThreaded/CPP/GlgViewer.cpp
#include "stdafx.h"
#include <stdio.h>
#include <stdlib.h>
#include "GlgViewer.h"

// Function prototype.
void OnTimerEvent( GlgViewer *viewer, GlgLong * timer_id );

/*----------------------------------------------------------------------
| Constructor. 
*/
GlgViewer::GlgViewer()
{
   UpdateInterval = 100; 
   TimerID = 0;
   TagRecordArray = NULL;
   NumTagRecords = 0;
   has_parent = false;
   DataFeed = NULL;

   // Set to False to use live data by default. 
   // It can be overriden by a command line option, 
   // -random-data or -live-data
   RANDOM_DATA = True; 
}

/*----------------------------------------------------------------------
| Destructor. 
*/
GlgViewer::~GlgViewer()
{
   // Clear TagRecordArray.
   DeleteTagRecords();
   
   // Delete DataFeed object, if defined. 
   if( DataFeed )
     delete DataFeed;
}

/*----------------------------------------------------------------------
| Add DataFeed object for supplying data for animation.
*/
void GlgViewer::AddDataFeed( DataFeedC * data_feed )
{
   if( DataFeed )
     delete DataFeed;
   
   DataFeed = data_feed;
}

/*----------------------------------------------------------------------
| Initialize drawing parameters as needed. Callbacks must be enabled
| before hierarchy setup.
*/
void GlgViewer::Init( void )
{
   /* Initialize drawing parameters before hierarchy is set up. */
   InitBeforeH();
   
   /* Set up object hierarchy. */ 
   SetupHierarchy();

   /* Initialize drawing parameters after hierarchy setup took place. */
   InitAfterH();
}

/*--------------------------------------------------------------------
| Place custom code as needed to initialize  the drawing before
| hierarchy setup took place.
*/
void GlgViewer::InitBeforeH()
{
   /* Callbacks must be enabled before  hierarchy setup.
      Enable callbacks only if the viewer object is a top level window and 
      there is no parent. 
      
      When this class is used in an MFC application, GlgSCADAViewer is 
      displayed in a GLG MFC control, has_parent=true and callbacks are 
      enabled at the parent level in GlgViewerControl::Init() before 
      the hierarchy setup.
   */
   if( !has_parent )
      EnableCallback( GLG_INPUT_CB, NULL );

   /* Place custom code here. */
}

/*--------------------------------------------------------------------
| Initialize the drawing after hierarchy setup. In this example, 
| it builds an array of tag records, which will be used to animate
| the drawing.
*/
void GlgViewer::InitAfterH()
{
   /* Store tag information as an array of records for faster processing */
   CreateTagRecords();
}

/*--------------------------------------------------------------------
| Obtains a list of tags in the loaded drawing and builds
| an array of tag records used to animate the drawing in a timer procedure
| UpdateDrawing().
*/
void GlgViewer::CreateTagRecords()
{
   CONST char 
     * tag_source,
     * tag_name,
     * tag_comment;
   GlgLong 
     i, 
     size;
   double dtype;
   double access_type;

   NumTagRecords = 0;
   
   /* Query a list of tags defined in the drawing */
   GlgObjectC tag_list = (GlgObjectC) CreateTagList( /* List each tag source only once */ True );

   /* CreateTagList creates an object (a group containing a list of tags).
      The returned object has to be explicitly dereferenced to prevent a 
      memory leak. The object is still referenced by the tag_list variable 
      instance.
   */
   tag_list.Drop();

   if( tag_list.IsNull() )
     return;   
   
   size = tag_list.GetSize();
   if( !size )
     return;   /* no tags found */
   
   TagRecordArray = (TagRecordC **) GlgAlloc( sizeof( TagRecordC * ) * size );

   for( i=0; i<size; ++i )
   {
      GlgObjectC tag_obj = (GlgObjectC) tag_list.GetElement( i );

      /* Get the name of the database field to use as a data source */
      tag_obj.GetResource( (char *) "TagSource", &tag_source );

      /* Obtain TagName and TagComment and use it as needed. */
      tag_obj.GetResource( "TagName", &tag_name );
      tag_obj.GetResource( "TagComment", &tag_comment );

      /* Skip tags with undefined TagSource */
      if( !tag_source || !*tag_source || ( strcmp( tag_source, "unset" ) == 0 ) )
	continue;

      /* Obtain tag access type. */
      tag_obj.GetResource( "TagAccessType", &access_type );

      /* Skip OUTPUT tags and INIT_ONLY tags. 
         Initialiaze INPUT_ONLY tags as needed. 
      */
      if( access_type == GLG_OUTPUT_TAG )
	continue;
      else if( access_type == GLG_INIT_ONLY_TAG )
      {
         InitTag( tag_source );
         continue;
      }

      /* Get tag object's data type: GLG_D, GLG_S or GLG_G */
      tag_obj.GetResource( "DataType", &dtype );

      /* Populate a new tag record in TagRecordArray. */
      /* Create a new tag record. */
      TagRecordC * tag_record = new TagRecordC();
      tag_record->tag_source = GlgStrClone( (char*) tag_source );
      tag_record->data_type = (int) dtype;
      tag_record->tag_obj = tag_obj;

      /* For further performance optimization, set if_changed=true which will
         push the value into the tag only if the value has changed.
         The if_changed flag is ignored for tags attached to the plots 
         in a real time chart, and the new value is always pushed to the 
         chart even if it is the same.
      */
      tag_record->if_changed = True;        

      // Store new tag_record in TagRecordArray.
      TagRecordArray[ NumTagRecords ] = tag_record;

      ++NumTagRecords;    /* Number of used tag records */
   }

   if( !NumTagRecords )
   {
      GlgFree( TagRecordArray );
      TagRecordArray = NULL;
   }
}

/*--------------------------------------------------------------------
| Clear TagRecordArray.
*/
void GlgViewer::DeleteTagRecords()
{
   GlgLong i;

   if( !NumTagRecords )
      return;

   /* Free memory for the tag_source */
   for( i = 0; i < NumTagRecords; ++i )
      delete TagRecordArray[i];
      
   /* Free memory allocated for the TagRecordArray */
   GlgFree( TagRecordArray );
   
   TagRecordArray = NULL;
   NumTagRecords = 0;
}

/*--------------------------------------------------------------------
| Initialize tag with a given tag source.
*/
void GlgViewer::InitTag( CONST char * tag_source )
{
   double value = -1.0;
   
   if( RANDOM_DATA )
   {
      /* Initialize "State" tag. */
      if( strcmp( tag_source, "State" ) == 0 )
        value = 1.0;
   }
   else;
   // Place custom code here to set value as needed.
   
   // Set the tag value in the drawing.
   SetTag( tag_source, value, False );
}


/*----------------------------------------------------------------------
| Timer procedure, invoked periodically on a timer to update the
| drawing with new data values.
*/
void OnTimerEvent( GlgViewer *viewer, GlgLong * timer_id )
{
   if( !viewer->NumTagRecords )
   {
      viewer->TimerID = 0;
      return;   // Don't restart the timer: no tags found.
   }

   // Update the drawing with new data values. 
   viewer->UpdateDrawing();

   /* Restart the timer. */
   viewer->TimerID = GlgAddTimeOut( viewer->AppContext, viewer->UpdateInterval, 
                                    (GlgTimerProc)OnTimerEvent, viewer );

}

/*--------------------------------------------------------------------
| Animate the drawing: 
| - Traverse TagRecordArray
| - For each tag record, obtain a new data value and push it into the graphics
| - Refresh the display.
*/
void GlgViewer::UpdateDrawing()
{
   TagRecordC * tag_record;
   double d_value;
   char * s_value;
   long i;   
   bool status;
   
   if( !NumTagRecords )
     return;

   if( !DataFeed )
   {
      GlgError( GLG_USER_ERROR, (char *) "Invalid DataFeed object, drawing animation failed." );
      return;
   }
   
   for( i=0; i<NumTagRecords; ++i )
   {
      tag_record = TagRecordArray[ i ];

      switch( tag_record->data_type )
      {
       case GLG_D:
         /* Obtain a new data value. */
         status = DataFeed->ReadDValue( tag_record, &d_value );
         if( !status )
         {
            /* Generate an error. */
            GlgError( GLG_USER_ERROR, (char *) "ReadDValue failed." );
            continue;
         }

         /* Push new value into graphics. */
         SetTag( tag_record->tag_source, d_value, 
                 tag_record->if_changed /*if_changed*/ );

         /* Optional performance optimization: faster */
         /* tag_record->tag_object.SetResource( NULL, d_value, tag_record->if_changed );
         */
         break;

       case GLG_S:         
         /* Obtain a new data value. */
         status = DataFeed->ReadSValue( tag_record, &s_value );
         if( !status )
         {
            /* Generate an error. */
            GlgError( GLG_USER_ERROR, (char *) "ReadSValue failed." );
            continue;
         }

         /* Push the new value into graphics. */
         SetTag( tag_record->tag_source, s_value, tag_record->if_changed /*if_changed*/ );

         /* Optional performance optimization: Faster */
         /* tag_record->tag_object.SetResourceIf( NULL, s_value, tag_record->if_changed ); 
         */
         break;
      }
   }

   /* Update the drawing with new data. */
   Update();
   Sync();    /* Improves interactive response. */
}

/*----------------------------------------------------------------------
| Handle user interaction. 
*/
void GlgViewer::Input( GlgObjectC& viewport, GlgObjectC& message )
{
   CONST char
     * format,
     * action,
     * origin;
      
   bool status = False;
   double d_value = -1.;;

   /* Get the message's format, action and origin. */
   message.GetResource( "Format", &format );
   message.GetResource( "Action", &action );
   message.GetResource( "Origin", &origin );

   /* Handle window closing. May use viewport's name. */
   if( strcmp( format, "Window" ) == 0 &&
      strcmp( action, "DeleteWindow" ) == 0 )
     exit( 0 );
   else if( strcmp( format, "Button" ) == 0 )
   {
      if(  strcmp( action, "Activate" ) !=0 && 
           strcmp( action, "ValueChanged" ) !=0 )
        return;

      if( strcmp( action, "Activate" ) == 0 ) /* Push button event */
      {
         if( strcmp( origin, "StartButton")  == 0 )
         {
            d_value = 1.;

            // Open the valve and start the pipe flow.
            if( RANDOM_DATA )
            {
               status = DataFeed->WriteDValue( "State", d_value );
               StartUpdates();
            }
            else
              // Place custom code here to write value to the appropriate tag.
              // status = DataFeed->WriteDValue( "State", d_value );
              ;
         }
         else if( strcmp( origin, "StopButton")  == 0 )
         {
            d_value = 0.;
            
            // Close the valve and stop the pipe flow.
            if( RANDOM_DATA )
            {
               status = DataFeed->WriteDValue( "State", d_value );
               StopUpdates();
            }
            else
              // Place custom code here to write value to the appropriate tag.
              // status = DataFeed->WriteDValue( "State", d_value );
              ;
         }
         else if( strcmp( origin, "QuitButton" ) == 0 )
         {
            exit( 0 );
         }
      }
      else if(  strcmp( action, "ValueChanged" ) == 0 ) /* Toggle button */
      {
         double state;
         message.GetResource( "OnState", &state );

         /* Place custom code here to handle toggle buttons as needed. */
      } 

      /* Refresh the display. */
      Update();
   } 
   else if( strcmp( format, "Timer" ) == 0 )
   {
      /* Refresh the display for the objects with integrated Timer dynamics. */
      Update();
   }
}


/*----------------------------------------------------------------------
| Start dynamic updates.
*/
void GlgViewer::StartUpdates()
{
   /* Start update timer. */
   TimerID = GlgAddTimeOut( AppContext, UpdateInterval, 
                           (GlgTimerProc)OnTimerEvent, this );
}


/*----------------------------------------------------------------------
| Stop dynamic updates.
*/
void GlgViewer::StopUpdates()
{
   /* Stop update timers */
   if( TimerID )
   {
      GlgRemoveTimeOut( TimerID );
      TimerID = 0;
   }
}

/*----------------------------------------------------------------------
| Set viewer size in screen cooridnates. 
*/
void GlgViewer::SetSize( GlgLong x, GlgLong y, 
                         GlgLong width, GlgLong height )
{
   SetResource( "Point1", 0., 0., 0. );
   SetResource( "Point2", 0., 0., 0. );

   SetResource( "Screen/XHint", (double) x );
   SetResource( "Screen/YHint", (double) y );
   SetResource( "Screen/WidthHint", (double) width );
   SetResource( "Screen/HeightHint", (double) height );
}

// Assignment operator
GlgViewer& GlgViewer::operator= ( const GlgObjectC& object )
{
   GlgObjectC::operator=( object );
   return *this;
}
